cql3:statement_restrictions.cc add more conditions to prevent "allow filtering" error

Jager227 opened the pull request:
To pop up in delete/update statements.

Fixes #12474

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

avikivity commented:
Please add a regression test (see test/cql-pytest)

nyh commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -443,7 +443,7 @@ statement_restrictions::statement_restrictions(data_dictionary::database db,
         if (_has_queriable_regular_index && _partition_range_is_simple) {
             _uses_secondary_indexing = true;
-        } else if (!allow_filtering) {
+        } else if (!allow_filtering && !type.is_delete() && !type.is_update()) {
```
This condition looks suspicious to me: I think this case should be an error even for DELETE and UPDATE, just a different error message. Or do we already get an error in that case from some other code path? Maybe it's all correct, but I can't know for sure without a good test designed to check exactly this. I think a good test should:
1. Run `DELETE FROM tbl WHERE partitionKey = ? AND value = ?` and verify we get an error, but not the wrong one suggesting ALLOW FILTERING would work. Just like in Cassandra. If this request doesn't fail and instead just ignores the `value = ?` condition, it's a mistake.
2. Maybe Run `DELETE FROM tbl WHERE partitionKey = ? AND value = ? ALLOW FILTERING` and check it still doesn't work (it's a syntax error).
3. Maybe also run `DELETE FROM tbl WHERE partitionKey = ? clusteringKey=?` and see it still works, we didn't break it.

Jager227 replied:
Results:
1.`InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid where clause contains non PRIMARY KEY columns: value"`
2.`SyntaxException: line 1:55 : Syntax error`
3. I think you missed "AND" between partitionKey = ? and clusteringKey = ?. If that is the case - this command passes without errors.

nyh replied:
1. Good, I didn't realize that we already had the "Invalid where clause contains non PRIMARY KEY columns: value" check somewhere else.
2. Good.
3. Yes.
So good.

Jager227 commented:
> > Please add a regression test (see test/cql-pytest)
> 
> The PR claims to fix #12474 which lists a bunch of translated Cassandra tests which demonstrate this issue. Sometimes Cassandra tests are large and fail for more than one reason, but please check which of these tests now passes (in which case the "xfail" mark should be removed), which continue to fail (in this case, please say why).
> 
> If there is no good test that verifies issue and only it, please add one, e.g. in test/cql-pytest/.

All 6 test currently fails, first 5 of them are failing with this error:
`AssertionError: Regex pattern did not match.
E        Regex: 'Non\\ PRIMARY\\ KEY\\ columns\\ found\\ in\\ where\\ clause:\\ value'
E        Input: 'Error from server: code=2200 [Invalid query] message="Invalid where clause contains non PRIMARY KEY columns: value"'`
It seems that these 2 messages are similar, so I can say that these tests can be considered passed
Test `cassandra_tests/validation/operations/update_test.py::testUpdateWithMultiplePartitionKeyComponents` failed with this error:
`AssertionError: Regex pattern did not match.
E        Regex: 'Some\\ partition\\ key\\ parts\\ are\\ missing:\\ partitionkey_2'
E        Input: 'Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"'`
Looks like this one still has some weird error output

nyh commented:
I think both phrasings of the error message are fine and we don't need to change ours to match Cassandra's exactly - instead I would change the 5 locations in the test to look just for the common string (e.g., perhaps "where clause" is a good enough string... or use a regex (add "RE" to the function name) and just allow the two variants. Then you can hopefully make the 5 tests pass and remove their "xfail" tag.

I don't know about the last one - if you think it is a new/separate bug, we should create a new issue explaining it, and the test should be modified to refer to the new issue in the "xfail" and/or comments - not to the old issue that will be closed when it is fixed.

Please also take a look at my comment above on something which looked like a wrong condition to me. I'm afraid we might not have a test that covers that case and need to add one.

nyh commented:
So to summarize, I think your patch is good, I just want to see what can be done about the various xfailing tests mentioned in https://github.com/scylladb/scylladb/issues/12474, as discussed above. The same pull-request should also fix and/or un-xfail these tests. Thanks.

Jager227 commented:
> So to summarize, I think your patch is good, I just want to see what can be done about the various xfailing tests mentioned in #12474, as discussed above. The same pull-request should also fix and/or un-xfail these tests. Thanks.

So, I managed to get 4 tests to pass but 1 of them has many checks for the content of the error message. The test's name is `cassandra_tests/validation/operations/delete_test.py::testDeleteWithSecondaryIndices`.
It has a whole wall of checks with different error messages, but we only got one returned (Invalid where clause contains non PRIMARY KEY columns: value).
`# Reproduces #12474:
        assertInvalidMessage(cql, table, "Non PRIMARY KEY columns found in where clause: value",
                             "DELETE FROM %s WHERE partitionKey = ? AND clustering_1 = ? AND value = ?", 3, 3, 3)
        assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS",
                             "DELETE FROM %s WHERE partitionKey = ? AND clustering_1 = ? AND values CONTAINS ?", 3, 3, 3)
        assertInvalidMessage(cql, table, "Non PRIMARY KEY columns found in where clause: value",
                             "DELETE FROM %s WHERE partitionKey = ? AND value = ?", 3, 3)
        assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS",
                             "DELETE FROM %s WHERE partitionKey = ? AND values CONTAINS ?", 3, 3)

        assertInvalidMessage(cql, table, "partitionkey",
                             "DELETE FROM %s WHERE clustering_1 = ?", 3)

        # Reproduces #12474:
        assertInvalidMessage(cql, table, "partitionkey",
                             "DELETE FROM %s WHERE value = ?", 3)`
 So the question is what should i do with this wall of checks?

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

Jager227 commented:
Rebased on latest master.

nyh commented:
> > So to summarize, I think your patch is good, I just want to see what can be done about the various xfailing tests mentioned in #12474, as discussed above. The same pull-request should also fix and/or un-xfail these tests. Thanks.
> 
> @nyh So, I managed to get 4 tests to pass

Thanks.

> but 1 of them has many checks for the content of the error message. The test's name is `cassandra_tests/validation/operations/delete_test.py::testDeleteWithSecondaryIndices`. It has a whole wall of checks with different error messages, but we only got one returned (Invalid where clause contains non PRIMARY KEY columns: value). assertInvalidMessage(cql, table, "Non PRIMARY KEY columns found in where clause: value", "DELETE FROM %s WHERE partitionKey = ? AND clustering_1 = ? AND value = ?", 3, 3, 3) assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS", "DELETE FROM %s WHERE partitionKey = ? AND clustering_1 = ? AND values CONTAINS ?", 3, 3, 3) assertInvalidMessage(cql, table, "Non PRIMARY KEY columns found in where clause: value", "DELETE FROM %s WHERE partitionKey = ? AND value = ?", 3, 3) assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS", "DELETE FROM %s WHERE partitionKey = ? AND values CONTAINS ?", 3, 3) assertInvalidMessage(cql, table, "partitionkey", "DELETE FROM %s WHERE clustering_1 = ?", 3) # Reproduces #12474: assertInvalidMessage(cql, table, "partitionkey", "DELETE FROM %s WHERE value = ?", 3) So the question is what should i do with lines that require other type of error message?

I'm afraid I don't understand the details. The main point is that after we mark issue #12474 fixes, we shouldn't leave any test with a `@pytest.mark.xfail(reason="#12474")` marker, or part of it (e.g., one test has `@pytest.mark.xfail(reason="#4244, #12474")`). If after fixing 12474 a test still fails, either the test should be change a bit to be more forgiving to error messages which are only slightly different, or, if we discover a new bug that was hiding while issue 12474 broke this test, then perhaps we need to open a new issue about it. We shouldn't leave a test claiming that it fails because of an issue that is already closed.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (11/21815):
- [test_not_created_compaction_task_abort](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/test_compaction_task/Tests___Unit_Tests___test_not_created_compaction_task_abort) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_not_created_compaction_task_abort)
- [cql.lwt_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.debug.1)
- [cql.lwt_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.debug.2)
- [cql.lwt_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.debug.3)
- [rest_api.test_compaction_task.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___rest_api_test_compaction_task_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+rest_api.test_compaction_task.debug.1)
- [cql.lwt_test.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.dev.1)
- [cql.lwt_test.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.dev.2)
- [cql.lwt_test.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.dev.3)
- [cql.lwt_test.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.release.1)
- [cql.lwt_test.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4606/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_lwt_test_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql.lwt_test.release.2)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4606/
- Duration: 7 hr 50 min
- Builder: i-0c3d802d6ef154460

Jager227 commented:
> I'm afraid I don't understand the details. The main point is that after we mark issue #12474 fixes, we shouldn't leave any test with a `@pytest.mark.xfail(reason="#12474")` marker, or part of it (e.g., one test has `@pytest.mark.xfail(reason="#4244, #12474")`). If after fixing 12474 a test still fails, either the test should be change a bit to be more forgiving to error messages which are only slightly different, or, if we discover a new bug that was hiding while issue 12474 broke this test, then perhaps we need to open a new issue about it. We shouldn't leave a test claiming that it fails because of an issue that is already closed.
 
Currently I have 2 questions. 
1) Is it ok if I remove assertions like `assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS", "DELETE FROM %s WHERE partitionKey = ? AND values CONTAINS ?", 3, 3)` from `cassandra_tests/validation/operations/delete_test.py::testDeleteWithSecondaryIndices`. The test fails because we get different type of error in this assertions.
2) As we discussed above test `cassandra_tests/validation/operations/update_test.py::testUpdateWithMultiplePartitionKeyComponents` returns partition key error which means that this can be another issue. The question is what should I do with this test now? If we will open a new issue, should I just remove pytest.mark.xfail(reason="#12474") or I need to wait for the issue to be created or I should create it myself and then put a new xfail mark?

Jager227 commented:
Rebased once again

nyh commented:
> Currently I have 2 questions.
> 
>     1. Is it ok if I remove assertions like `assertInvalidMessage(cql, table, "Cannot use DELETE with CONTAINS", "DELETE FROM %s WHERE partitionKey = ? AND values CONTAINS ?", 3, 3)` from `cassandra_tests/validation/operations/delete_test.py::testDeleteWithSecondaryIndices`. The test fails because we get different type of error in this assertions.

In this case I usually added a comment and replaced the assertInvalidMessage by assertInvalid which doesn't check the message, e.g., in update_test.py I did:
```python
        # Cassandra complains about "Cannot use UPDATE with CONTAINS, but
        # Scylla complains about "Cannot use CONTAINS on non-collection".
        assertInvalid(cql, table,
                             "UPDATE %s SET value = ? WHERE partitionKey CONTAINS ? AND clustering_1 = ?", 7, 0, 1)
```
Another, perhaps even better, option is to use assertInvalidMessageRE (the regular-expression version) and allow both versions - the Scylla one and Cassandra one - as reasonable outputs.


>     2. As we discussed above, test `cassandra_tests/validation/operations/update_test.py::testUpdateWithMultiplePartitionKeyComponents` returns partition key error which means that this can be another issue. The question is what should I do with this test now? If we will open a new issue, should I just remove `pytest.mark.xfail(reason="#12474")` or I need to wait for the issue to be created or I should create it myself and then put a new xfail mark?

The error you saw, suggesting ALLOW FILTERING, seems like the same bug we are claiming to fix here, isn't it? So why didn't our patch fix it? Maybe our patch is missing a fix somewhere?

But - we need to check if this is a real bug in Scylla or just something missing in the test. Please try the following: If the error message tells you an "ALLOW FILTERING" is missing, try adding ALLOW FILTERING to the request in the test. If the request still fails, then this is a real bug - we should suggest ALLOW FILTERING in cases it doesn't help (this is #12474). If the request starts to pass, then Scylla's error message is **correct**, and the test should be modified to allow either Cassandra's or Scylla's message (again with assertInvalidMessageRE()).

Jager227 commented:
Good day, @nyh. It seems that `lwt_test.cql` currently fails, because it expects "allow filtering" error in update query, but receives "Invalid where clause contains non PRIMARY KEY columns: b". In this patch we removed allow filtering error from update queries, so this test must be modified too, right?

Query: `UPDATE lwt SET b = 2 WHERE b = 1`
Expected error: `Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"`
Result:`Error from server: code=2200 [Invalid query] message="Invalid where clause contains non PRIMARY KEY columns: b"`

Jager227 commented:
By the way, all related Cassandra tests now pass. I'll push when I deal with lwt_test.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

Jager227 commented:
rebased

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5035/)
:white_check_mark: - [Sanity Tests](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6391/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5035/)
#### Build Details:
- Duration: 3 hr 47 min
- Builder: i-096d6559bbd9c255a (m5d.12xlarge)

avikivity commented:
@nyh please complete the review

nyh commented:
> @nyh please complete the review

Thanks. I'm trying now to clear my huge review backlog, without much success (hi @kostja :-)). I'll get to review this one tomorrow morning. Sorry about the delay.

nyh commented:
> Rebased on latest master

I don't know why github continues to complain about "This branch has conflicts that must be resolved". Maybe another rebase is needed?

Jager227 commented:
Rebased and pushed without changes, still got conflict...

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

denesb commented:
Please don't merge master into your branch. Instead, rebase on top of master.

Jager227 commented:
> Please don't merge master into your branch. Instead, rebase on top of master.

Sorry, i'll try to rebase again

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

Jager227 commented:
I fixed my branch, but the conflict is still present.

denesb commented:
> I fixed my branch, but the conflict is still present. I also rebased it, but git says that current branch it up to date

Note that the conflict might be with `next`, you might need another rebase after promotion.

nyh commented:
> > I fixed my branch, but the conflict is still present. I also rebased it, but git says that current branch it up to date
> 
> Note that the conflict might be with `next`, you might need another rebase after promotion.

But his PR's base is "master", not next, so why would github complain about a conflict with next? Strange.
Anyway, when I'm ready to commit this patch, I'll see for myself if there's any conflict or not, I don't care what github thinks :-)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5096/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6455/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5096/)
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

Jager227 commented:
>     2. Please check that no test remains with an **xfail** whose reason is 12474. If I count correctly, I think two such tests remained in delete_test.py (please correct me if I'm wrong), and the 12474 should be removed from their xfail (or if you discovered a new bug, please open a new issue for it and mention it). After this patch is merged, issue 12474 will be closed - and we shouldn't have tests claiming they are known to fail because of it.

Yes you were right, there were 2 tests with `xfail#12474`, I removed this tags and modified them as well
I also returned `Reproduces:` comments and fixed some grammar in my other comments in this patch

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5099/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6461/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5099/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: spider8.cloudius-systems.com

nyh commented:
The typo-fixing patch reached master. So please rebase on the current master now. Thanks.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15811

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5122/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6497/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5122/)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider3.cloudius-systems.com

nyh commented:
I'll merge, but I see that Cassandra tests don't work smoothly now, for example:
```
test/cql-pytest/run-cassandra cassandra_tests/validation/operations/compact_storage_test.py
```
fails two tests. I *think* this is not the fault of your patch (I think the failing test is different from what you modified?), but we'll have to check later and see what happened.
